snuba/query/snql/parser.py

```diff
@@ -58,48 +72,52 @@
     visit_parameters_list,
     visit_quoted_literal,
 )
+from snuba.util import parse_datetime
 
 snql_grammar = Grammar(
     r"""
-    query_exp             = match_clause where_clause? collect_clause? group_by_clause? having_clause? order_by_clause? limit_clause?
-
-    match_clause          = space* "MATCH" space* entity_match space*
-    where_clause          = space* "WHERE" or_expression space*
-    collect_clause        = space* "COLLECT" collect_list space*
-    group_by_clause       = space* "BY" group_list space*
-    having_clause         = space* "HAVING" or_expression space*
-    order_by_clause       = space* "ORDER BY" order_list space*
-    limit_clause          = space* "LIMIT" space* integer_literal space*
+    query_exp             = match_clause select_clause group_by_clause? where_clause? having_clause? order_by_clause? limit_by_clause? limit_clause? offset_clause? granularity_clause? totals_clause? space*
+
+    match_clause          = space* "MATCH" space+ entity_match
+    select_clause         = space+ "SELECT" space+ select_list
+    group_by_clause       = space+ "BY" space+ group_list
+    where_clause          = space+ "WHERE" space+ or_expression
+    having_clause         = space+ "HAVING" space+ or_expression
+    order_by_clause       = space+ "ORDER BY" space+ order_list
+    limit_by_clause       = space+ "LIMIT" space+ integer_literal space* "BY" space* column_name
+    limit_clause          = space+ "LIMIT" space+ integer_literal
+    offset_clause         = space+ "OFFSET" space+ integer_literal
+    granularity_clause    = space+ "GRANULARITY" space+ integer_literal
+    totals_clause         = space+ "TOTALS" space+ boolean_literal
+    sample_clause         = space+ "SAMPLE" space+ numeric_literal
+
+    entity_match          = open_paren entity_alias colon space* entity_name sample_clause? space* close_paren
+
+    and_expression        = space* condition (space* and_tuple)*
+    or_expression         = space* and_expression (space* or_tuple)*
+    and_tuple             = space* "AND" condition
+    or_tuple              = space* "OR" and_expression
```

Comment:
shouldn't those spaces be mandatory ?

Reply:
Yes they should be, but changing the * to a + breaks a number of the tests. Honestly trying to change the parser from optional spaces to required spaces has been a nightmare. I would rather chip away at this then make this PR even longer.

Reply:
Isn't that because there is `space*` before `or_tuple` and `and_tuple` on lines 97 and 98, which is already consuming all the spaces. Why not just remove these 2 additional `space*`? (i.e. revert lines 99 and 100 to the way it was before)